client/lib/browser-storage/bypass.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * External dependencies
+ */
+import debugModule from 'debug';
+
+const debug = debugModule( 'calypso:support-user' );
+
+// This module defines a series of methods which bypasse all persistent storage.
+// Any calls to read/write data using browser-storage instead access a temporary
+// in-memory store which is lost on page reload. This driver is used to sandbox
+// a user's data while support-user is active, ensuring it does not contaminate
+// the original user, and vice versa.
+
+const memoryStore: { [index: string]: any } = {};
+
+export function getStoredItem< T >( key: string ): Promise< T | null > {
+	debug( 'browser-storage bypass', 'getStoredItem', key );
+
+	if ( key in memoryStore ) {
+		// Match localforage behaviour: unknown keys return null.
+		return Promise.resolve( memoryStore[ key ] !== undefined ? memoryStore[ key ] : null );
+	}
+
+	return Promise.resolve( null );
+}
+
+export function setStoredItem< T >( key: string, value: T ): Promise< void > {
+	debug( 'browser-storage bypass', 'setStoredItem', key );
+
+	return new Promise( resolve => {
+		memoryStore[ key ] = value !== undefined ? value : null;
+		resolve();
+	} );
```

Comment:
I opted for the former, to avoid triggering our overly-prescriptive lint rule that `async` functions should have at least one `await` (should we disable that?)

Reply:
The lint rule was added here: https://github.com/Automattic/wp-calypso/pull/30697

I suspect the rule helps prevents some misuse of `async`, but I don't feel strongly.

Feel free to open a PR to propose the removal or disable the lint rule with a comment for your valid use case.

Reply:
With lint rules it's all a balance between helping you catch more errors before they're pushed, or just getting in the way of the code you want to write. And that's highly dependent on the developer, their experience with that particular feature, and the project's general guidelines of how to write code.

It's not clear to me whether this rule is causing more harm or good. It helps avoid some obvious errors, like forgetting to remove the `async` keyword when e.g. refactoring an async function to sync, but it some terse code like the one @jsnajdr was suggesting. Is that code something that comes up often? Is it readable and clear?

It seems that adding this particular rule was a conscious decision at some point, though (not just something that slipped in with a bunch of other rules), so perhaps we should leave it be? 🤷‍♂

Reply:
I think we found a reasonable use case where the ESLint rule is an obstacle: implementing an interface that specifies async functions (i.e., Promise return values), but not doing anything async inside the implementation:
```js
class StorageWithLocalStorage extends Storage {
  async set( key, value ) {
    localStorage.setItem( key, value );
  }
  async get( key ) {
    return localStorage.getItem( key );
  }
}
```
Let's remove the rule from our config. I vaguely remember that the conversation with @sirreal who added it was like: look, here's an interesting new rule in the latest ESLint, what if we try it out? 🙂

Reply:
I'll prepare a PR, then!

Reply:
That's #34032, for anyone following this thread.

Reply:
I simplified things as we were discussing, after rebasing on top of master and picking up the changes in #34032.